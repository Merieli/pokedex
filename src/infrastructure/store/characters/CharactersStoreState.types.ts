export interface CharactersStoreState {
    characters: any[];
}
