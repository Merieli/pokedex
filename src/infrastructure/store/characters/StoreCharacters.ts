import { defineStore } from 'pinia';
import { CharactersStoreState } from './CharactersStoreState.types';
import getAllCharacters from './actions/getAllCharacters';

const state = (): CharactersStoreState => ({
    characters: [],
});

const getters = {};

const actions = {
    getAllCharacters,
};

export const useCounterStore = defineStore('characters', {
    state,
    actions,
});
