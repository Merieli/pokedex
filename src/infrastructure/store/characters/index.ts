export * from './CharactersStoreState.types';
export * from './StoreCharacters';
