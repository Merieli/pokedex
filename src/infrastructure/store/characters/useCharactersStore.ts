import { Ref, ref } from 'vue';

import { Character } from '@domain/index';
import { defineStore } from 'pinia';

import { logError } from '@/helpers/logError';
import CharactersGatewayHttp from '@/infrastructure/gateway/CharactersGatewayHttp';

export const useCharactersStore = defineStore('characters', () => {
    const listOfCharacters: Ref<Character[]> = ref([]);
    const isLoading: Ref<boolean> = ref(false);

    const getAllCharacters = async (): Promise<void> => {
        try {
            isLoading.value = true;
            const api = new CharactersGatewayHttp();
            const response = await api.getAll();
            listOfCharacters.value = response.results;
        } catch (error) {
            const { code } = logError('Action getAllCharacters', error);
            // notifyUser(code, 'imprimir todos os personagens')  //função para exibir o erro para o usuário
        } finally {
            isLoading.value = false;
        }
    };

    return {
        listOfCharacters,
        isLoading,
        getAllCharacters,
    };
});
