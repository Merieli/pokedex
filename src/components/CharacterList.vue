<script setup lang="ts">
import { ComputedRef, computed, onMounted } from 'vue';

import { useQuantityBasedOnWidth } from '@composables/useQuantityBasedOnWidth';
import CardLoading from './CardLoading.vue';
import CharacterCard from './CharacterCard.vue';

import { useSearchActions } from '@/composables/useSearchActions';
import { useSelectedCharacter } from '@/composables/useSelectedCharacter';
import { Character } from '@/domain';
import { router } from '@/infrastructure/router/router';
import { useCharactersStore } from '@/infrastructure/store/characters';

const store = useCharactersStore();
const { totalOfCardsLoading, totalOfPagination } = useQuantityBasedOnWidth();
const { clearSearch } = useSearchActions();
const { setsSelectedCharacter } = useSelectedCharacter();

onMounted(async () => {
    await store.getAllCharacters();
});

const listOfCharacters: ComputedRef<Character[]> = computed(() => {
    if (store.random.show && !store.isSearching && !store.isLoading) return store.random.character;

    if (store.isSearching && !store.isLoading) return store.search.characters;

    if (!store.isLoading) return store.charactersPerPage[store.pagination.currentPage];

    return [];
});

const changeThePage = async (page: number) => {
    store.$patch({
        isLoading: true,
        pagination: { currentPage: page },
    });

    if (!store.charactersPerPage[store.pagination.currentPage]) {
        await store.getAllCharacters();
    }
    store.$patch({
        isLoading: false,
    });
};

const openCharacter = (character: Character) => {
    router.push({ name: 'character', params: { id: character.id } });

    setsSelectedCharacter(character);
};
</script>

<template>
    <section class="character-list">
        <h4
            v-if="store.random.show && !store.isSearching && !store.isLoading"
            class="character-list__text"
            data-list="clear-random"
        >
            Clear character and show list
            <v-btn
                density="comfortable"
                size="small"
                variant="tonal"
                class="character-list__button-clear"
                icon="mdi-close"
                data-list="clear-random-button"
                @click="clearSearch"
            ></v-btn>
        </h4>

        <h4 v-else-if="store.isSearching && !store.isLoading" class="character-list__text" data-list="results-search">
            1 - {{ store.search.pagination.total }} of more than {{ store.search.pagination.results }} results to "{{
                store.search.text
            }}". Clear search and show list
            <v-btn
                density="comfortable"
                size="small"
                variant="tonal"
                class="character-list__button-clear"
                icon="mdi-close"
                data-list="clear-search-button"
                @click="clearSearch"
                @keyup.enter="clearSearch"
            ></v-btn>
        </h4>

        <h4 v-else-if="!store.isLoading" class="character-list__text" data-list="results">
            {{ store.pagination.results }} results
        </h4>

        <main class="character-list__cards">
            <CharacterCard
                v-for="(character, index) in listOfCharacters"
                :id="Number(character.id)"
                :key="index"
                :name="character.name"
                :species="character.species"
                :image="character.image"
                :alt-image="`${character.name} of specie ${character.species}`"
                data-list="card"
                @open="openCharacter(character)"
            />
            <div v-if="store.isLoading" class="character-list__loading">
                <CardLoading :quantity="totalOfCardsLoading" />
            </div>
        </main>

        <v-pagination
            v-if="store.pagination.total > 0 && !store.isSearching && !store.random.show"
            v-model="store.pagination.currentPage"
            :length="store.pagination.total"
            :total-visible="totalOfPagination"
            rounded="circle"
            data-list="pagination"
            variant="elevated"
            active-color="meri-light"
            @update:model-value="changeThePage"
        ></v-pagination>
    </section>
</template>

<style lang="postcss" scoped>
.character-list {
    @apply px-10 mb-[230px]
        w-full max-w-5xl
        grid place-items-center grid-cols-1;

    &__text {
        @apply mt-4 mb-6 p-4
            flex justify-between
            font-semibold text-lg text-gray-500 text-left
            rounded-xl
            w-full;
    }

    &__button-clear {
        @apply ml-4;
    }

    &__cards {
        @apply py-2 
            flex flex-row flex-wrap
            w-full;
    }

    &__loading {
        @apply flex flex-row flex-wrap
            mt-12 w-full h-full;
    }
}
</style>

<style>
.v-btn--disabled.v-btn--variant-elevated .v-btn__content {
    font-weight: 700;
    color: gray;
}

.v-pagination__prev .v-btn--disabled.v-btn--variant-elevated,
.v-pagination__item .v-btn--disabled.v-btn--variant-elevated {
    background-color: white;
    color: white;
}

.v-pagination__prev .v-btn--disabled.v-btn--variant-elevated .v-btn__content,
.v-pagination__item .v-btn--disabled.v-btn--variant-elevated .v-btn__content {
    color: theme('colors.gray.300');
}
</style>
