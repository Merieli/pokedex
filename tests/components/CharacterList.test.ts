import { nextTick } from 'vue';

import CharacterList from '@components/CharacterList.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { createVuetify } from 'vuetify';
import { mockListOfCharactersInSearch, mockStoreCharacters } from './__mocks__/mockStoreCharacters';

import { useCharactersStore } from '@/infrastructure/store/characters';

describe('CharacterList.vue', () => {
    const setupWrapper = () => {
        const pinia = createTestingPinia({
            initialState: {
                characters: {
                    pagination: {
                        currentPage: 1,
                        total: 1,
                        results: 10,
                    },
                    charactersPerPage: mockStoreCharacters,
                },
            },
        });
        const vuetify = createVuetify();

        return {
            wrapper: mount(CharacterList, {
                global: {
                    plugins: [pinia, vuetify],
                },
            }),
            store: useCharactersStore(),
        };
    };

    describe('Integração/Componente', () => {
        describe('👀 Renderização:', () => {
            test('Dado o componente Quando renderizado Então deve possuir html equivalente ao Snapshot gravado', () => {
                const { wrapper } = setupWrapper();

                expect(wrapper.html()).toMatchSnapshot();
            });

            test('Dado uma lista de personagens Quando renderizada Então deve mostrar a quantidade total de resultados encontrados em todas as páginas', () => {
                const { wrapper } = setupWrapper();
                const qtd = wrapper.find('[data-list="results"]');
                const totalOfCharacters = 10;

                expect(qtd.text()).toBe(`${totalOfCharacters} results`);
            });

            test('Dado uma lista de personagens Quando renderizada Então deve mostrar o card de personagens na quantidade correta', () => {
                const { wrapper } = setupWrapper();
                const qtd = wrapper.findAll('[data-list="card"]');

                expect(qtd).toHaveLength(4);
            });

            test('Dado uma lista de personagens Quando renderizada Então deve exibir um botão para carregar mais personagens', () => {
                const { wrapper } = setupWrapper();
                const loadPage = wrapper.findAll('[data-list="pagination"]');

                expect(loadPage).toHaveLength(1);
            });
        });
        describe('🧠 Comportamento:', () => {
            test('Dado a lista de personagens Quando montado o componente Então deve chamar a action para buscar todos os pesonagens', () => {
                const { wrapper, store } = setupWrapper();

                expect(store.getAllCharacters).toHaveBeenCalledTimes(1);
            });

            test('Dado uma paginação Quando clicada em uma página Então deve chamar a action para buscar mais personagens da próxima página', async () => {
                const { wrapper, store } = setupWrapper();
                const loadPage = wrapper.find('[data-list="pagination"]');

                await loadPage.trigger('click');

                expect(store.getAllCharacters).toHaveBeenCalledTimes(1);
            });

            test('Dado uma lista de personagens Quando buscar por algum personagem em específico Então deve trocar os personagens visíveis pelos encontrados na store', async () => {
                const { wrapper, store } = setupWrapper();
                store.search.characters = mockListOfCharactersInSearch;
                await nextTick();
                const cards = wrapper.findAll('[data-list="card"]');
                const names = wrapper.findAll('[data-character="name"]');

                expect(cards).toHaveLength(2);
                expect(names[0].text()).toBe(mockListOfCharactersInSearch[0].name);
                expect(names[1].text()).toBe(mockListOfCharactersInSearch[1].name);
            });
        });
    });
});
