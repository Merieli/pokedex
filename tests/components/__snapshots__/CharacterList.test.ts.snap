// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterList.vue > Integra√ß√£o/Componente > üëÄ Renderiza√ß√£o: > Dado o componente Quando renderizado Ent√£o deve possuir html equivalente ao Snapshot gravado 1`] = `
"<section class=\\"character-list\\" data-v-c3db22c7=\\"\\">
  <h4 class=\\"character-list__qtd\\" data-list=\\"qtd\\" data-v-c3db22c7=\\"\\">1 Characters</h4>
  <main class=\\"character-list__cards\\" data-v-c3db22c7=\\"\\">
    <section class=\\"character-card\\" data-character=\\"card\\" data-list=\\"card\\" data-v-82d09848=\\"\\" data-v-c3db22c7=\\"\\"><img class=\\"character-card__image\\" data-character=\\"image\\" src=\\"https://image.com/rick.png\\" alt=\\"character Rick teste\\" data-v-82d09848=\\"\\">
      <p class=\\"character-card__id\\" data-character=\\"id\\" data-v-82d09848=\\"\\">#1</p>
      <h2 class=\\"character-card__name\\" data-character=\\"name\\" data-v-82d09848=\\"\\">Rick teste</h2>
      <div class=\\"character-card__tag\\" data-character=\\"species\\" style=\\"background-color: rgb(242, 163, 139);\\" data-v-82d09848=\\"\\">Human</div>
    </section>
  </main>
  <v-btn-stub class=\\"character-list__button\\" variant=\\"tonal\\" data-list=\\"button-load-more\\" data-v-c3db22c7=\\"\\"></v-btn-stub>
</section>"
`;
